File: packages/python/elizaos/bootstrap/actions/remove_contact.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.bootstrap.utils.xml import parse_key_value_xml
from elizaos.types import (
    Action,
    ActionExample,
    ActionResult,
    Content,
    ModelType,
)

if TYPE_CHECKING:
    from elizaos.types import (
        HandlerCallback,
        HandlerOptions,
        IAgentRuntime,
        Memory,
        State,
    )


REMOVE_CONTACT_TEMPLATE = """# Remove Contact from Rolodex

Current message: {{message}}
Sender: {{senderName}} (ID: {{senderId}})

Extract removal information:
1. Who to remove (name or entity reference)
2. Confirmation (yes or no)

<response>
<contactName>Name of the contact to remove</contactName>
<confirmed>yes or no</confirmed>
</response>"""


@dataclass
class RemoveContactAction:
    name: str = "REMOVE_CONTACT"
    similes: list[str] = field(
        default_factory=lambda: [
            "DELETE_CONTACT",
            "REMOVE_FROM_ROLODEX",
            "DELETE_FROM_CONTACTS",
        ]
    )
    description: str = "Removes a contact from the rolodex"

    async def validate(
        self, runtime: IAgentRuntime, _message: Memory, _state: State | None = None
    ) -> bool:
        rolodex_service = runtime.get_service("rolodex")
        return rolodex_service is not None

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        from elizaos.bootstrap.services.rolodex import RolodexService

        rolodex_service = runtime.get_service("rolodex")
        if not rolodex_service or not isinstance(rolodex_service, RolodexService):
            return ActionResult(
                text="Rolodex service not available",
                success=False,
                values={"error": True},
                data={"error": "RolodexService not available"},
            )

        state = await runtime.compose_state(message, ["RECENT_MESSAGES", "ENTITIES"])

        prompt = runtime.compose_prompt_from_state(
            state=state,
            template=REMOVE_CONTACT_TEMPLATE,
        )

        response = await runtime.use_model(ModelType.TEXT_SMALL, {"prompt": prompt})
        parsed = parse_key_value_xml(response)

        if not parsed or not parsed.get("contactName"):
            return ActionResult(
                text="Could not determine which contact to remove",
                success=False,
                values={"error": True},
                data={"error": "No contact name provided"},
            )

        contact_name = str(parsed.get("contactName", ""))
        confirmed = str(parsed.get("confirmed", "no")).lower() == "yes"

        if not confirmed:
            response_text = (
                f'To remove {contact_name}, please confirm by saying "yes, remove {contact_name}".'
            )
            if callback:
                await callback(Content(text=response_text, actions=["REMOVE_CONTACT"]))
            return ActionResult(
                text=response_text,
                success=True,
                values={"needsConfirmation": True},
                data={"contactName": contact_name},
            )

        contacts = await rolodex_service.search_contacts(search_term=contact_name)

        if not contacts:
            return ActionResult(
                text=f"Could not find a contact named '{contact_name}'",
                success=False,
                values={"error": True},
                data={"error": "Contact not found"},
            )

        contact = contacts[0]
        removed = await rolodex_service.remove_contact(contact.entity_id)

        if removed:
            response_text = f"I've removed {contact_name} from your contacts."
            if callback:
                await callback(Content(text=response_text, actions=["REMOVE_CONTACT"]))
            return ActionResult(
                text=response_text,
                success=True,
                values={"contactId": str(contact.entity_id)},
                data={"success": True},
            )
        else:
            return ActionResult(
                text="Failed to remove contact",
                success=False,
                values={"error": True},
                data={"error": "Remove operation failed"},
            )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return [
            [
                ActionExample(
                    name="{{name1}}", content=Content(text="Remove John Doe from my contacts")
                ),
                ActionExample(
                    name="{{name2}}",
                    content=Content(
                        text='To remove John Doe, please confirm by saying "yes, remove John Doe".',
                        actions=["REMOVE_CONTACT"],
                    ),
                ),
            ],
        ]


remove_contact_action = Action(
    name=RemoveContactAction.name,
    similes=RemoveContactAction().similes,
    description=RemoveContactAction.description,
    validate=RemoveContactAction().validate,
    handler=RemoveContactAction().handler,
    examples=RemoveContactAction().examples,
)

File: plugins/plugin-rolodex/src/actions/removeContact.ts
import {
  type Action,
  type IAgentRuntime,
  type Memory,
  type State,
  logger,
  parseKeyValueXml,
  composePromptFromState,
  type HandlerCallback,
  ModelType,
  type ActionResult,
} from '@elizaos/core';
import { RolodexService } from '../services/RolodexService';

const removeContactTemplate = `# Remove Contact from Rolodex

Current message: {{message}}
Sender: {{senderName}} (ID: {{senderId}})

## Instructions
Extract the contact removal information from the message:
1. Who to remove (name or entity reference)
2. Confirmation of the intent to remove

## Response Format
<response>
<contactName>Name of the contact to remove</contactName>
<confirmed>yes or no</confirmed>
</response>`;

export const removeContactAction: Action = {
  name: 'REMOVE_CONTACT',
  similes: ['DELETE_CONTACT', 'REMOVE_FROM_ROLODEX', 'DELETE_FROM_CONTACTS'],
  description: 'Removes a contact from the rolodex',

  validate: async (runtime: IAgentRuntime, message: Memory, state?: State): Promise<boolean> => {
    const hasService = !!runtime.getService('rolodex');
    const hasIntent = message.content.text
      ?.toLowerCase()
      .match(/remove|delete|drop.*contact|remove.*from.*rolodex/);
    return hasService && !!hasIntent;
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    _options?: { [key: string]: unknown },
    callback?: HandlerCallback
  ): Promise<ActionResult | void> => {
    try {
      const rolodexService = runtime.getService('rolodex') as RolodexService;
      if (!rolodexService) {
        throw new Error('RolodexService not available');
      }

      // Compose the prompt
      const removeState = {
        ...state,
        message: message.content.text,
        senderName: state?.senderName || 'User',
        senderId: message.entityId,
      };

      const prompt = composePromptFromState({
        state: removeState as State,
        template: removeContactTemplate,
      });

      // Get LLM response
      const response = await runtime.useModel(ModelType.TEXT_SMALL, { prompt });
      const parsed = parseKeyValueXml(response);

      if (!parsed?.contactName) {
        logger.warn('[RemoveContact] No contact name provided');
        await callback?.({
          text: "I couldn't determine which contact to remove. Please specify the contact name.",
        });
        return;
      }

      if (parsed.confirmed !== 'yes') {
        await callback?.({
          text: `To remove ${parsed.contactName} from your contacts, please confirm by saying "yes, remove ${parsed.contactName}".`,
        });
        return;
      }

      // Find the contact
      const contacts = await rolodexService.searchContacts({ searchTerm: parsed.contactName });

      if (contacts.length === 0) {
        await callback?.({
          text: `I couldn't find a contact named "${parsed.contactName}" in the rolodex.`,
        });
        return;
      }

      const contact = contacts[0];

      // Remove the contact
      const removed = await rolodexService.removeContact(contact.entityId);

      if (removed) {
        const responseText = `I've removed ${parsed.contactName} from your contacts.`;
        await callback?.({
          text: responseText,
          actions: ['REMOVE_CONTACT'],
        });

        logger.info(`[RemoveContact] Removed contact ${contact.entityId}`);

        return {
          success: true,
          values: { contactId: contact.entityId },
          data: { success: true },
          text: responseText,
        };
      } else {
        throw new Error('Failed to remove contact');
      }
    } catch (error) {
      logger.error('[RemoveContact] Error:', error instanceof Error ? error.message : String(error));
      await callback?.({
        text: 'I encountered an error while removing the contact. Please try again.',
        error: error instanceof Error ? error.message : 'Unknown error',
      });
    }
  },

  examples: [
    [
      {
        name: '{{name1}}',
        content: {
          text: 'Remove John Doe from my contacts',
        },
      },
      {
        name: '{{name2}}',
        content: {
          text: 'To remove John Doe from your contacts, please confirm by saying "yes, remove John Doe".',
          actions: ['REMOVE_CONTACT'],
        },
      },
    ],
    [
      {
        name: '{{name1}}',
        content: {
          text: 'Yes, remove John Doe',
        },
      },
      {
        name: '{{name2}}',
        content: {
          text: "I've removed John Doe from your contacts.",
          actions: ['REMOVE_CONTACT'],
        },
      },
    ],
  ],
};
